apps/passport-client/components/modals/AccountImportModal.tsx

```diff
@@ -0,0 +1,253 @@
+import { SemaphoreIdentityPCDTypeName } from "@pcd/semaphore-identity-pcd";
+import { useCallback, useEffect, useState } from "react";
+import styled from "styled-components";
+import { useFilePicker } from "use-file-picker";
+import { useDispatch, useModal, usePCDCollection } from "../../src/appHooks";
+import { getPackages } from "../../src/pcdPackages";
+import { Button } from "../core";
+
+// There are three main UI states that can occur after a user selects a file
+// to import.
+type ImportState =
+  | { state: "ready" }
+  // The selected file is valid, and the user can decide whether to import it.
+  | {
+      state: "valid-file-selected";
+      collection: PCDCollection;
+      pcdsToMergeIds: Set<PCD["id"]>;
+    }
+  // The import has been carried out, and `added` is the number of PCDs
+  // imported.
+  | { state: "import-complete" }
+  // The selected file is not valid.
+  | { state: "invalid-file" };
+
+export function AccountImportModal() {
+  const { openFilePicker, filesContent } = useFilePicker({
+    accept: ".json",
+    multiple: false,
+    onFilesSelected: () => {
+      setImportState({ state: "ready" });
+    }
+  });
+
+  const modal = useModal();
+  const pcdCollection = usePCDCollection();
+  const dispatch = useDispatch();
+
+  const [importState, setImportState] = useState<ImportState>({
+    state: "ready"
+  });
+
+  // Called when a valid file has been selected, and the user chooses to import
+  // PCDs from it.
+  const importPCDs = useCallback(() => {
+    // Should never happen, but makes TypeScript happy that we checked for it
+    if (importState.state !== "valid-file-selected") return;
+
+    dispatch({
+      type: "merge-import",
+      collection: importState.collection,
+      pcdsToMergeIds: importState.pcdsToMergeIds
+    });
+
+    setImportState({ state: "import-complete" });
+  }, [dispatch, importState]);
+
+  // Responds to the user having selected a file to import, or to changes in
+  // the user's current PCD collection.
+  useEffect(() => {
+    (async () => {
+      // If a file has been selected, and isn't invalid or already imported
+      if (
+        filesContent.length > 0 &&
+        importState.state !== "import-complete" &&
+        importState.state !== "invalid-file"
+      ) {
+        let importedCollection: PCDCollection;
+
+        // If the file hasn't been processed yet, process it
+        if (importState.state === "ready") {
+          try {
+            // Parse the file content as JSON
+            const storageExport = JSON.parse(filesContent[0].content);
+            // Deserialize the storage - throws an error if the content is not
+            // recognized
+            const importedBackup = await deserializeStorage(
+              storageExport,
+              await getPackages()
+            );
+            importedCollection = importedBackup.pcds;
+            if (!(importedCollection instanceof PCDCollection)) {
```

Comment:
Is this really checking for the wrong type, or checking for undefined/null?
It doesn't hurt, but I think seeing the wrong type here would indicate deserializeStorage was violating its return type, which the compiler should rule out.

Reply:
Ok. I can't remember exactly why I thought this was necessary, so I'll remove it.